programs/ziti-edge-tunnel/include/log_utils.c

```diff
@@ -0,0 +1,315 @@
+        case 4:
+            err_level = "DEBUG";
+            break;
+        case 5:
+            err_level = "VERBOSE";
+            break;
+        case 6:
+            err_level = "TRACE";
+            break;
+        default:
+            err_level = "UNKNOWN";
+    }
+    return err_level;
+}
+
+void ziti_log_writer(int level, const char *loc, const char *msg, size_t msglen) {
+    char curr_time[32];
+    uv_timeval64_t now;
+    uv_gettimeofday(&now);
+    struct tm *tm = gmtime(&now.tv_sec);
+
+    snprintf(curr_time, sizeof(curr_time), "%04d-%02d-%02dT%02d:%02d:%02d.%03dZ",
+             1900 + tm->tm_year, tm->tm_mon + 1, tm->tm_mday,
+             tm->tm_hour, tm->tm_min, tm->tm_sec, now.tv_usec / 1000
+    );
+
+    if ( ziti_tunneler_log != NULL) {
+        fputc('\n', ziti_tunneler_log);
+        fprintf(ziti_tunneler_log, "[%s] %7s %s ", curr_time, parse_level(level), loc);
+        fwrite(msg, 1, msglen, ziti_tunneler_log);
+    }
+
+    if(multi_writer) {
+        printf("\n[%s] %7s %s %.*s", curr_time, parse_level(level), loc, msglen, msg);
+    }
+
+}
+
+bool open_log(char* log_filename) {
+    if((ziti_tunneler_log=fopen(log_filename,"a")) == NULL) {
+        printf("Could not open logs file %s", log_filename);
+        return false;
+    }
+    return true;
+}
+
+void close_log() {
+    if (ziti_tunneler_log != NULL) {
+        fclose(ziti_tunneler_log);
+        ziti_tunneler_log = NULL;
+    }
+    if (log_filename != NULL) {
+        free(log_filename);
+        log_filename = NULL;
+    }
+}
+
+void stop_log_check() {
+    uv_check_stop(log_flusher);
+}
+
+void rotate_log() {
+    close_log();
+
+    uv_timeval64_t file_time;
+    uv_gettimeofday(&file_time);
+    struct tm* orig_time = gmtime(&file_time.tv_sec);
+    memcpy(start_time, orig_time, sizeof(struct tm));
+    log_filename = get_log_filename();
+
+    open_log(log_filename);
+}
+
+void delete_older_logs(uv_loop_t *ziti_loop) {
+    char curr_path[FILENAME_MAX]; //create string buffer to hold path
+#if _WIN32
+    GetCurrentDir( curr_path, FILENAME_MAX );
+#else
+    sprintf(curr_path, "%s", "/tmp");
+#endif
+
+    char log_path[FILENAME_MAX];
+    sprintf(log_path, "%s/logs", curr_path);
+
+    uv_fs_t fs;
+    int rc = uv_fs_scandir(ziti_loop, &fs, log_path, 0, NULL);
+    if (rc < 7) {
+        if (rc < 0) {
+            ZITI_LOG(ERROR, "failed to scan dir[%s]: %d/%s", log_path, rc, uv_strerror(rc));
+        } else {
+            ZITI_LOG(TRACE, "Log files count in [%s] is %d, not deleting log files.", log_path, rc);
+            uv_fs_req_cleanup(&fs);
+        }
+        return;
+    }
+
+    char **log_files = calloc(rc + 1 , sizeof(char *));
+    uv_dirent_t file;
+    int rotation_cnt = 0;
+    while (uv_fs_scandir_next(&fs, &file) == 0) {
+        ZITI_LOG(TRACE, "log file = %s %d", file.name, file.type);
+
+        if (file.type == UV_DIRENT_FILE) {
+            if (memcmp(file.name, log_filename_base, sizeof(log_filename_base)) == 0) {
```

Comment:
log_filename_base is a `const char *`. `sizeof(const char *)` is not the same as the length of string.

but more crucially, `memcmp` does not stop comparing characters when it encounters a '\0' in either buffer. You really want `strncmp` here instead of `memcmp`, and the length limit of the comparison should be `strlen(log_filename_base)` instead of sizeof.